### ISSUE TITLE
Length requiremments for DB table columns

### ISSUE THREAD

**Issue by lookitsjonno** [defuse/php-encryption#89]
Hello,

I am encoding my ciphertext with Crypto::BinToHex($ciphertext) ready for storage in the database. Is there a way to find out the maximum length that the ciphertext can be? Previously the column that I want to encrypt was a VARCHAR(40).

It would be good if the documentation was updated with a formula to calculate this, otherwise ciphertexts will be cut short.

Jonathan.

**Comment by paragonie-scott** [comment]
This is going to change if my proposal for 2.0.0 is accepted. Use `TEXT` or `MEDIUMTEXT` if you want an easy answer.

1. The message in raw binary is:
    * 32 bytes MAC
    * 16 bytes IV
    * the message, which is always a multiple of 16 bytes in length
2. Hex encoding doubles the output length.

The best way to figure this out is to run the following code:

```php
$str = Crypto::binToHex(Crypto::encrypt($old_column, $key));
var_dump(strlen($str));
```

In 2.0.0, there will be a little bit more overhead:
1. A salt will be included in the HKDF calculation (16 bytes)
2. A version tag will be included on all ciphertext (4 bytes)

This additional overhead comes with two advantages:
1. Much lower chance of an `(key, iv)` pair repeating
2. Forward compatibility, should the library's primitives need to be tweaked

TL;DR you're looking at a minimum of 64 bytes (or 128 hex characters), which will eventually be a minimum of 84 bytes (168 hex character).

**Comment by lookitsjonno** [comment]
Thanks for the quick reply, it's refreshing to find an encryption library that is easy to use that follows best practise! :+1:
